File: app/src/main/java/i/am/shiro/amai/widget/SearchInput.java
package i.am.shiro.amai.widget;

import android.content.Context;
import android.graphics.Rect;
import android.text.Editable;
import android.util.AttributeSet;
import android.view.KeyEvent;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;

import com.annimon.stream.Optional;
import com.annimon.stream.function.Consumer;
import com.google.android.material.textfield.TextInputEditText;

public final class SearchInput extends TextInputEditText {

    private Consumer<String> onSubmitListener;

    private Consumer<String> onTextChangedListener;

    public SearchInput(Context context, AttributeSet attrs) {
        super(context, attrs);
        setImeOptions(EditorInfo.IME_ACTION_SEARCH);
    }

    public void setOnSubmitListener(Consumer<String> onSubmitListener) {
        this.onSubmitListener = onSubmitListener;
    }

    public void setOnTextChangedListener(Consumer<String> onTextChangedListener) {
        this.onTextChangedListener = onTextChangedListener;
    }

    @Override
    public boolean onKeyPreIme(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            dismiss();
            return true;
        }

        return super.onKeyPreIme(keyCode, event);
    }

    @Override
    public void onEditorAction(int actionCode) {
        if (actionCode == EditorInfo.IME_ACTION_SEARCH) {
            dismiss();
            Optional.ofNullable(getText())
                    .map(Editable::toString)
                    .ifPresent(onSubmitListener);
        }

        super.onEditorAction(actionCode);
    }

    @Override
    protected void onFocusChanged(boolean focused, int direction, Rect previouslyFocusedRect) {
        if (!focused) {
            hideKeyboard();
        }

        super.onFocusChanged(focused, direction, previouslyFocusedRect);
    }

    @Override
    protected void onTextChanged(CharSequence text, int start, int lengthBefore, int lengthAfter) {
        if (onTextChangedListener != null) onTextChangedListener.accept(text.toString());
    }

    private void dismiss() {
        hideKeyboard();
        clearFocus();
    }

    private void hideKeyboard() {
        InputMethodManager imm = (InputMethodManager) getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(getWindowToken(), 0);
    }
}

File: app/src/main/java/i/am/shiro/amai/widget/SearchInput.kt
package i.am.shiro.amai.widget

import android.content.Context
import android.graphics.Rect
import android.util.AttributeSet
import android.view.KeyEvent
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import com.google.android.material.textfield.TextInputEditText
import timber.log.Timber

class SearchInput(context: Context, attrs: AttributeSet) : TextInputEditText(context, attrs) {

    var onSubmitListener: (String) -> Unit = {}

    // nullable necessary
    var onTextChangedListener: ((String) -> Unit)? = null

    init {
        Timber.wtf("RABBIT 123")
        imeOptions = EditorInfo.IME_ACTION_SEARCH
    }

    override fun onKeyPreIme(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            dismiss()
            return true
        }

        return super.onKeyPreIme(keyCode, event)
    }

    override fun onEditorAction(actionCode: Int) {
        if (actionCode == EditorInfo.IME_ACTION_SEARCH) {
            dismiss()
            text?.toString()?.let { onSubmitListener(it) }
        }

        super.onEditorAction(actionCode)
    }

    override fun onFocusChanged(focused: Boolean, direction: Int, previouslyFocusedRect: Rect?) {
        if (!focused) {
            hideKeyboard()
        }

        super.onFocusChanged(focused, direction, previouslyFocusedRect)
    }

    override fun onTextChanged(text: CharSequence, start: Int, lengthBefore: Int, lengthAfter: Int) {
        onTextChangedListener?.invoke(text.toString())
    }

    private fun dismiss() {
        hideKeyboard()
        clearFocus()
    }

    private fun hideKeyboard() {
        val imm = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(windowToken, 0)
    }
}
